middleware/router.go

```diff
@@ -141,6 +143,143 @@ func DefaultRouter(spec *loads.Document, api RoutableAPI) Router {
 	return builder.Build()
 }
 
+// RouteAuthenticator is an authenticator that can compose several authenticators together.
+// It also knows when it contains an authenticator that allows for anonymous pass through.
+// Contains a group of 1 or more authenticators that have a logical AND relationship
+type RouteAuthenticator struct {
+	Authenticator  map[string]runtime.Authenticator
+	Schemes        []string
+	Scopes         map[string][]string
+	allScopes      []string
+	commonScopes   []string
+	allowAnonymous bool
+}
+
+func (ra *RouteAuthenticator) AllowsAnonymous() bool {
+	return ra.allowAnonymous
+}
+
+// AllScopes returns a list of unique scopes that is the combination
+// of all the scopes in the requirements
+func (ra *RouteAuthenticator) AllScopes() []string {
+	return ra.allScopes
+}
+
+// CommonScopes returns a list of unique scopes that are common in all the
+// scopes in the requirements
+func (ra *RouteAuthenticator) CommonScopes() []string {
```

Comment:
I've made both of the choices available, but it's unclear to me what you behavior you'd want if you have 2 oauth auts defined as AND.
if you have oauth A with scopes: users:admin, posts:admin
and oauth B with scopes: categories:admin, posts:admin
 
I've chosen a union here as default so that the scopes resulting scopes here will be: users:admin, posts:admin, categories:admin

Now if you want to get the result: posts:admin, categories:admin instead you can do:
```go
middelware.MatchedRouteFor(params.HTTPRequest).Authenticator.CommonScopes()
```

Reply:
yeah this is what I figured out from the code.
the troublesome point is that swagger "officially" supports scopes only for oauth.
I haven't tried it yet but I am almost sure that go-swagger wouldn't mind scopes in say, an API key header.

This way, I may analyze _several_ headers and intersect or cross the resolved scopes (I actually have both cases: I intend the result to be passed back to the operation via the principal structure.

I'll give you an example application, when some business transaction is intermediated:
- to post an "order", I need to identify the end buyer and the intermediary
- the end buyer must satisfy some claims in its token, and so does the intermediary
- in the end, the operation may be carried on only if both satisfy these claims